### PR TITLE
[elasticsearch] Keystore integration

### PR DESCRIPTION
**Pull request by Crazybus**
- [x] Chart version *not* bumped (the versions are all bumped and released at the same time)
- [x] README.md updated with any new values or changes
- [x] Updated template tests in `${CHART}/tests/*.py` 
- [x] Updated integration tests in `${CHART}/examples/*/test/goss.yaml`

### CODE REVIEW

File: elasticsearch/README.md [hunk 1]
```diff
@@ -171,18 +172,9 @@ There are a couple reasons we recommend this.
...
-   ```
-2. Mount it into the container via `secretMounts`
-   ```
-   secretMounts:
-   - name: elasticsearch-keystore
-     secretName: elasticsearch-keystore
-     path: /usr/share/elasticsearch/config/elasticsearch.keystore
-     subPath: elasticsearch.keystore
-   ```
+Take a look a the [config example](/elasticsearch/examples/config/values.yaml) which has a tested version of adding strings and files to the keystore.
```
Comment (jordansissel) [inline elasticsearch/README.md]: Might be helpful to have a note explicitly stating that any external values (secrets, etc) need to be explicitly set as env vars (which is what the example does). Thoughts?

File: elasticsearch/examples/config/values.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - name: elastic-config-secret
+    secretName: elastic-config-secret
+    path: /usr/share/elasticsearch/config/secret
+
+keystore:
+  enabled: true
+  strings:
+    bootstrap.password: '${ELASTIC_PASSWORD}'
+    xpack.notification.slack.account.monitoring.secure_url: '${SLACK_URL}'
+  files:
```
Comment (jordansissel) [inline elasticsearch/examples/config/values.yaml]: cloud-on-k8s doesn't have a specific concept of file vs string keystore entries: https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-es-secure-settings.html

I don't recall why that is, but I wonder if we can have the same concepts for this on both the chart and the operator?

Comment (Crazybus) [reply elasticsearch/examples/config/values.yaml]: I remember a change was made (initiated by the cloud-on-k8s team) to have the Elasticsearch keystore treat all strings and files the same internally. 

My original assumption was that the operator would be doing something special to take these secrets, generate a keystore and create that as a secret. But their approach is also [fully init container based](https://github.com/elastic/cloud-on-k8s/blob/0f5aab786ea99f1374dceca1e86fca2c385de117/operators/pkg/controller/common/keystore/initcontainer.go#L33-L51). All secrets are mounted as files and then the init script loops over them and adds them to the keystore. 

I like this a lot more than my solution, mainly because you don't need to think about mapping the files and strings into the container yourself. And the other huge security advantage is that these secrets won't still exist as environment variables or mounts in the actual running Elasticsearch container. 

One downside to this approach is that it looks like the spec only allows a single secret. The downside of this being that you now need to fully recreate your keystore secret everytime you want to add something new. 

Making it a list would solve this so I'll ping the team and see what their thoughts on it are. 
```
spec:
  secureSettings:
    - secretName: your-secure-settings-secret
    - secretName: other-secure-settings-secret
```

I'm going to take a stab at implementing this in the same way.

**Comment by Crazybus** [comment]
@jordansissel I have updated it to match how cloud-on-k8s is doing it (which is a much better approach). Issues have also been opened to implement the suggestions I had around allowing multiple secrets, and the ability to set all parts of the volumeMount spec to allow specifying a different path than the key of the secret.

https://github.com/elastic/cloud-on-k8s/issues/1457
https://github.com/elastic/cloud-on-k8s/issues/1458

Please take another look :)

File: elasticsearch/templates/statefulset.yaml [hunk 3]
```diff
@@ -111,6 +111,14 @@ spec:
           configMap:
             name: {{ template "uname" . }}-config
         {{- end }}
+{{ if .Values.keystore }}
+        - name: keystore
+          emptyDir: {}
+{{ end }}
+        {{- range .Values.keystore }}
```
Comment (jordansissel) [inline elasticsearch/templates/statefulset.yaml]: curious, why not have this `{{- range ...}}` inside the `{{ if ... ` block above? I I don't think there'd be any behavior changes, just curious about the style choice.

File: elasticsearch/templates/statefulset.yaml [hunk 4]
```diff
@@ -111,6 +111,14 @@ spec:
           configMap:
             name: {{ template "uname" . }}-config
         {{- end }}
+{{ if .Values.keystore }}
```
Comment (jordansissel) [inline elasticsearch/templates/statefulset.yaml]: Did you want to indent this and use `{{- if ...` ? Not sure if we're concerned with indentation style.

File: elasticsearch/templates/statefulset.yaml [hunk 5]
```diff
@@ -129,6 +137,38 @@ spec:
...
+          elasticsearch-keystore create
+
+          for i in /tmp/keystoreSecrets/*/*; do
+            key=$(basename $i)
+            echo "Adding file $i to keystore key $key"
+            elasticsearch-keystore add-file "$key" "$i"
+          done
+
+          # Add the bootstrap password since otherwise the Elasticsearch entrypoint tries to do this on startup
+          [ ! -z "$ELASTIC_PASSWORD" ] && echo $ELASTIC_PASSWORD | elasticsearch-keystore add -x bootstrap.password
```
Comment (jordansissel) [inline elasticsearch/templates/statefulset.yaml]: Needs quotation marks around the second var, `echo "$ELASTIC_PASSWORD"` otherwise a password with whitespace won't be set correctly.

Comment (jordansissel) [reply elasticsearch/templates/statefulset.yaml]: It might be too complicated to add to the tests, but maybe we could add a test to cover this case (whitespace in a password, such as `ELASTIC_PASSWORD="hello       world     "`

Comment (Crazybus) [reply elasticsearch/templates/statefulset.yaml]: Great catch! Fixed! I thought I had copy pasted this from the ELasticsearch docker file, but looks like my copy paste skills failed since it is properly quoted in there: https://github.com/elastic/elasticsearch/blob/31f6e78acdb8dfc7c042be4ad0bc0bdea003e055/distribution/docker/src/docker/bin/docker-entrypoint.sh#L88

File: elasticsearch/templates/statefulset.yaml (hunk 4 above)
Comment (Crazybus) [reply elasticsearch/templates/statefulset.yaml]: Fixed! It would be nice to add a "white space detection" check in the pull request testing and do a one off PR to get things back in line. Right now its a bit of a mess. It's a shame that `helm lint --strict` doesn't already cover this.

File: elasticsearch/templates/statefulset.yaml (hunk 3 above)
Comment (Crazybus) [reply elasticsearch/templates/statefulset.yaml]: This is just a leftover from the original refactoring. But you are right that it doesn't change anything. I fixed it up to make it as explicit as possible.

**Comment by Crazybus** [comment]
Thanks for the review @jordansissel, all comments have been addressed. Please take another look :)

**Comment by Crazybus** [comment]
:heart:

And thank you for the great suggestions and finding bugs!

**Comment by dcvtruong** [comment]
@Crazybus What is the correct usage for "keystore: []" in values.yaml? I try several different format for '[]' format and it was not working.

keystore: 
   xpack.notification.slack.account.monitoring.secure_url: ''https://webhook_url"

**Comment by Crazybus** [comment]
The simpliest example is:

https://github.com/elastic/helm-charts/blob/a49995539e4347e8729437cfbb54299023c81306/elasticsearch/examples/config/values.yaml#L25-L26

With the secret created like this:

https://github.com/elastic/helm-charts/blob/a49995539e4347e8729437cfbb54299023c81306/elasticsearch/examples/config/Makefile#L12

**Comment by dcvtruong** [comment]
@Crazybus Thanks.

**Comment by harsh4870** [comment]
i have tried adding GCS service account key in keystone but getting error

created kubernetes secret and added file into it and added secret into keystore

but pod are crashing also my i am using custom docker image with plugin installed but getting error like : 

```
csearch", "node.name": "elasticsearch-master-0",  "message": "uncaught exception in thread [main]" ,
"stacktrace": ["org.elasticsearch.bootstrap.StartupException: java.lang.IllegalArgumentException: unknown secure setting [gcs_backup_key.json] please check that any required plugins are installed, or check the breaking changes documentation for removed settings",
"at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:163) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:150) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124) ~[elasticsearch-cli-7.3.2.jar:7.3.2]",
"at org.elasticsearch.cli.Command.main(Command.java:90) ~[elasticsearch-cli-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:115) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92) ~[elasticsearch-7.3.2.jar:7.3.2]",
"Caused by: java.lang.IllegalArgumentException: unknown secure setting [gcs_backup_key.json] please check that any required plugins are installed, or check the breaking changes documentation for removed settings",
"at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:531) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:476) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:447) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:418) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.common.settings.SettingsModule.<init>(SettingsModule.java:149) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.node.Node.<init>(Node.java:357) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.node.Node.<init>(Node.java:258) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:221) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:221) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:349) ~[elasticsearch-7.3.2.jar:7.3.2]",
"at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:159) ~[elasticsearch-7.3.2.jar:7.3.2]",
"... 6 more"] }
```
created secret using 

`kubectl create secret generic gcs-backup-key --from-file=gcs_backup_key.json=gcs_backup_key.json`

i want to set cronjob for setting regular backup let me know if missing anything thanks in advance for help.
